File: Design pattern/Builder.py
from abc import ABCMeta, abstractmethod

class Product:
    def __init__(self):
        self.m_parts = list()

    def print_parts(self):
        for ele in self.m_parts:
            print(ele)

class IBuilder(metaclass=ABCMeta):
    @abstractmethod
    def produce_part_a(self):
        pass

    @abstractmethod
    def produce_part_b(self):
        pass

    @abstractmethod
    def produce_part_c(self):
        pass

class ConcreteBuilder(IBuilder):
    def __init__(self):
        self.m_produce = Product()

    def produce_part_a(self):
        self.m_produce.m_parts.append("Part A")

    def produce_part_b(self):
        self.m_produce.m_parts.append("Part B")

    def produce_part_c(self):
        self.m_produce.m_parts.append("Part C")

class Director:
    def __init__(self):
        self.m_builder = None

    def build_minimal_featured_product(self):
        self.m_builder.produce_part_a()

    def build_full_featured_product(self):
        self.m_builder.produce_part_a()
        self.m_builder.produce_part_b()
        self.m_builder.produce_part_c()

if __name__ == "__main__":
    director = Director()

    builder = ConcreteBuilder()
    director.m_builder = builder
    director.build_minimal_featured_product()
    product = builder.m_produce
    product.print_parts()
    print("\n")

    builder = ConcreteBuilder()
    director.m_builder = builder
    director.build_full_featured_product()
    product = builder.m_produce
    product.print_parts()
    print("\n")

File: Design pattern/Builder.cpp
#include <iostream>
#include <vector>
using namespace std;

class Product
{
public:
    void print_parts() const
    {
        for (size_t i = 0; i < m_parts.size(); i++)
        {
            cout << m_parts[i] << endl;
        }
    }

public:
    vector<string> m_parts;
};

class IBuilder
{
public:
    virtual ~IBuilder() {}
    virtual void produce_part_a() = 0;
    virtual void produce_part_b() = 0;
    virtual void produce_part_c() = 0;
};

class ConcreteBuilder : public IBuilder
{
public:
    ConcreteBuilder() { m_product = new Product(); }
    ~ConcreteBuilder() { delete m_product; }

    void produce_part_a() override { m_product->m_parts.push_back("Part A"); }
    void produce_part_b() override { m_product->m_parts.push_back("Part B"); }
    void produce_part_c() override { m_product->m_parts.push_back("Part C"); }

public:
    Product *m_product;
};

class Director
{
public:
    void set_builder(IBuilder *builder) { m_builder = builder; }
    void build_minimal_featured_product() const
    {
        m_builder->produce_part_a();
    }
    void build_full_featured_product() const
    {
        m_builder->produce_part_a();
        m_builder->produce_part_b();
        m_builder->produce_part_c();
    }

private:
    IBuilder *m_builder;
};

int main()
{
    Director *dir;
    IBuilder *builder;
    Product *product;

    dir = new Director();

    builder = new ConcreteBuilder();
    product = static_cast<ConcreteBuilder*>(builder)->m_product;
    dir->set_builder(builder);
    dir->build_minimal_featured_product();
    product->print_parts();
    cout<<endl;
    delete builder;

    builder = new ConcreteBuilder();
    product = static_cast<ConcreteBuilder*>(builder)->m_product;
    dir->set_builder(builder);
    dir->build_full_featured_product();
    product->print_parts();
    cout<<endl;
    delete builder;

    delete dir;
}